libcudacxx/include/cuda/__memory/address_space.h

```
@@ -54,24 +54,121 @@ enum class address_space
   _CCCL_ASSERT(__ptr != nullptr, "invalid pointer");
   _CCCL_ASSERT(_CUDA_DEVICE::__cccl_is_valid_address_space(__space), "invalid address space");
 
+  // NVCC and NVRTC < 12.3 have problems tracking the address space of pointers, fallback to inline PTX for them
   switch (__space)
   {
     case address_space::global:
+#  if _CCCL_CUDA_COMPILER(NVCC, <, 12, 3) || _CCCL_CUDA_COMPILER(NVRTC, <, 12, 3)
+    {
+      unsigned __ret;
+      asm volatile(
+        "{\n\t"
+        "  .reg .pred p;\n\t"
+        "  isspacep.global p, %1;\n\t"
```

I am completely out of the loop here - I see no tracking issue so apologies if my questions seem naive.

Is `isspacep.global/shared/constant et al.` faster than the intrinsics `isGlobal/Shared/Constant et al.`? If yes, why? If no, what is the purpose of hardcoding these facilities for `nvcc/nvrtc > 12.3`?

---

The thing is that cuda compilers track pointer's address spaces and when using the intrinsics, some of the checks can be optimized out.

On the other hand if there are bugs in the compiler, it may lead to invalid results, which is what @miscco was dealing with in the past.

So we prefer the intrinsics and fallback to hardcoded PTX for buggy compilers :)